core/src/banking_stage/scheduler_messages.rs

```diff
@@ -0,0 +1,76 @@
+use {
+    crate::immutable_deserialized_packet::ImmutableDeserializedPacket,
+    solana_sdk::transaction::SanitizedTransaction, std::sync::Arc,
+};
+
+/// A unique identifier for a transaction batch.
+/// The lower 64 bits are based on the order the transaction batch was scheduled.
+/// The upper 64 bits are an optional priority field - currently unused.
+#[derive(Clone, Copy, Debug, Default, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TransactionBatchId(u128);
+
+impl TransactionBatchId {
+    pub fn new(index: u64) -> Self {
+        Self(index as u128)
+    }
+}
+
+/// A unique identifier for a transaction.
+/// The lower 64 bits are based on the order the transaction entered banking stage.
+/// The upper 64 bits are the priority of the transaction.
+#[derive(Clone, Copy, Debug, Default, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TransactionId(u128);
+
+impl TransactionId {
+    pub fn new(priority: u64, index: u64) -> Self {
+        Self(((priority as u128) << 64) | (index as u128))
+    }
+}
+
+/// Message: [Scheduler -> Worker]
+/// Transactions to be consumed (executed, recorded, committed)
+pub struct ConsumeWork {
+    pub batch_id: TransactionBatchId,
+    pub transaction_ids: Vec<TransactionId>,
+    pub transactions: Vec<SanitizedTransaction>,
+}
+
+/// Message: [Worker -> Scheduler]
+/// Transactions to be forwarded to the next leader(s)
+pub struct ForwardWork {
+    pub ids: Vec<TransactionId>,
+    pub packets: Vec<Arc<ImmutableDeserializedPacket>>,
+}
+
+/// Message: [Worker -> Scheduler]
+/// Processed transactions.
+pub struct FinishedConsumeWork {
+    pub work: ConsumeWork,
```

Comment:
does this make copy of `Vec<SanitizedTransaction>` and send back, results 2X transactions being send back-n-forth between scheduler and worker? If so, that could become a hot spot, given number of transactions should only increase in the future.

Reply:
We only need references to the transactions to process them ([relevant worker code](https://github.com/solana-labs/solana/pull/30970/files#diff-92c8957903dfca99b6396d081e155b6d939e1f952389149a52a2e5f47d47845eR112-R123)), so there's no need to take them out of the `ConsumeWork` struct

In the implementation (linked above), the `work` field here, is the original `ConsumeWork` received by the worker, no **cloning**, which I think is what you are concerned about?

To be entirely clear/pedantic, it will do a **copy** of the struct when we move it into the `Finished*` wrappers. But that's just a shallow-copy of the vecs, so the entire struct copy is only 64 bytes. We could reduce that to 8 bytes by wrapping it in a `Box`, but then we have 2 pointer follows to get to the txs; not sure it'd be worth it. @ryoqun might have some thoughts here 😄

Reply:
🤔 

So after writing the above comment, I was thinking "how can we enforce the fact the txs aren't cloned out by worker".

Maybe it's a better interface to not `pub` the fields, and add some impls for the messages:
1. `fn new() -> Self;`
2. `fn transactions(&self) -> &[SanitizedTransaction];`
3. `fn finish(retryable_indexes: Vec<usize>) -> FinishConsumeWork;`

and similarly for the `ForwardWork` message pair.

@taozhu-chicago, @ryoqun - do you think this is a better interface than what I've got here with the `pub` fields?

Reply:
>  the work field here, is the original ConsumeWork received by the worker, no cloning, which I think is what you are concerned about?

Yep, without see the actual implementation, I wasn't able to tell is `work: Vec<SanitizedTransaction>` is moved or copied. So yeah, would be a good idea to make message interface enforce `ref` to avoid accidental copy.

Reply:
🤔 
Thinking the interface I described above won't be sufficient to enforce that, since we need to get the values out on the finished side. Will think about a better design tonight.


Also back to my original comment, it's probably slightly misleading. We do have to copy (move not clone - so still not copying the underlying Vecs within the tx), the memory of the `SanitizedTransaction` from the Scheduler's storage into the Vec when we create the batch. And back from Vec into Scheduler storage iff it is returned as retryable.

Reply:
Struggling to come up with a good way to enforce this in rust. We still want the fields of `FinishedConsumeWork` to be pub so we can pull them out at the end, but also want to restrict construction, to come from an existing `ConsumeWork`.

Potentially something like this:
```rust
// needed so clippy doesn't yell about using #[non_exhaustive],
// which only restricts OUTSIDE the crate...
#[allow(clippy::manual_non_exhaustive)] 
pub struct FinishedConsumeWork {
    pub work: ConsumeWork,
    pub retryable_indexes: Vec<usize>,
    _priv: (),
}

impl FinishedConsumeWork {
    fn new(work: ConsumeWork, retryable_indexes: Vec<usize>) -> Self {
        Self {
            work,
            retryable_indexes,
            _priv: (),
        }
    }
}
```

but I think that's fairly ugly on the receiving side. 



After considering it more, I'm hesitant to put any implementation in what are supposed to be simple messages. wdyt @taozhu-chicago / @ryoqun ?

Reply:
yea that does feel like bit of over engineering, at least at this stage. I am cool with committing what's ready to move forward.

Reply:
yeah, i think current form is fine as-is.